src/compas/datastructures/assembly/assembly.py

```diff
@@ -2,68 +2,145 @@
+            Note that the key is unique only in the context of the current assembly.
+            Nested assemblies may have the same ``key`` value for one of their parts.
+            Default is ``None`` in which case the key will be automatically assigned integer value.
+        kwargs: dict
+            Additional named parameters colledted in a dict.
+
+        Returns
+        -------
+        int or str
+            The identifier of the part in the current assembly graph.
+
+        """
+        key = self.graph.add_node(key=key, part=part, **kwargs)
+        part.key = key
+        return key
+
+    def add_connection(self, a, b, **kwargs):
+        """Add a connection between two parts.
+
+        Parameters
+        ----------
+        a: int or str
+            The identifier of the "from" part in the current assembly.
+        b: int or str
+            The identifier of the "to" part in the current assembly.
+        kwargs: dict
+            Additional named parameters colledted in a dict.
+
+        Returns
+        -------
+        tuple of str or int
+            The tuple of node identifiers that identifies the connection.
+
+        """
+        return self.graph.add_edge(a, b, **kwargs)
 
     def parts(self):
-        """The parts of the assembly."""
-        return self.network.nodes()
+        """The parts of the assembly.
+
+        Yields
+        ------
+        :class:`compas.datastructures.Node`
+            The individual parts or sub-assemblies of the current assembly.
+        """
+        for node in self.graph.nodes():
+            yield self.graph.node_attribute(node, 'part')
 
     def connections(self):
         """The connections between the parts."""
-        return self.network.edges()
+        return self.graph.edges()
 
-    def add_part(self, part, id=None, **kwargs):
-        """Add a part to the assembly."""
-        self.network.add_node(key=id, part=part, **kwargs)
+    def find(self, guid):
+        """Find a part in the assembly.
 
-    def add_connection(self, a, b, **kwargs):
-        """Add a connection between two parts."""
-        self.network.add_edge(a, b, **kwargs)
+        This method will traverse all parts and sub-assemblies of the current assembly
+        to find the part.
+
+        Parameters
+        ----------
+        guid: str
+            A globally unique identifier.
+            This identifier is automatically assigned when parts are created.
+
+        Returns
+        -------
+        :class:`compas.datastructures.Node` or None
+            The identified part, if any.
+
+        """
+        for node in self.graph.nodes():
+            part = self.graph.node_attribute(node, 'part')
+            if part.guid == guid:
+                return part
+        # this is effectively a depth-first search through all the parts of the assembly
+        for node in self.graph.nodes():
+            part = self.graph.node_attribute(node, 'part')
+            part = part.find(guid)
```

Comment:
This could result in an infinite loop, but I don't know if we want to check for that or expect the user to not do that.

Reply:
Also, `Part`/`Node`s don't have a `find` method, so I think this would raise some exceptions.